### PR TITLE
Grant & Retrieve via Polygon Payments; Pre-release cleanup.

### PR DESCRIPTION
**Pull request by KPrasch**
**Type of PR:**
Bugfix & Feature

**Required reviews:** 
2

**What this does:**

- Polygon CLI and configuration updates
- Mitigations for network update metadata pollution
- Handles SSL certificates as part of RestMiddleware with just-in-time cert fetching
- Forcibly clear all stored node metadata and certificates on startup
- Only store peers in memory during runtime
- Working demos for polygon testnet
- Relax invalid node logging and errors
- Bumps probationary period three months
- Minor code quality improvements (type hints and organization)
- `NUCYPHER_WORKER_ETH_PASSWORD` -> `NUCYPHER_OPERATOR_ETH_PASSWORD`

### CODE REVIEW

**Comment by derekpierre** [comment]
@KPrasch what's needed to get this off of WIP?

**Comment by KPrasch** [comment]
First off the dependencies were updated which resulted in new test failures so we'll need to resolve that.  Ideally this can be followed up with the complete relocation of contract tests to nucypher-contracts to lessen the impact.

File: nucypher/network/nodes.py [hunk 1]
```diff
@@ -805,15 +801,15 @@ def learn_from_teacher_node(self, eager=False, canceller=None):
         # These except clauses apply to the current_teacher itself, not the learned-about nodes.
         except NodeSeemsToBeDown as e:
             unresponsive_nodes.add(current_teacher)
-            self.log.info(
-                f"Teacher {str(current_teacher)} is perhaps down:{e}.")  # FIXME: This was printing the node bytestring. Is this really necessary?  #1712
+            raise
```
Comment (derekpierre) [inline nucypher/network/nodes.py:804]: Was the `raise` here intended?

File: nucypher/network/nodes.py [hunk 2]
```diff
@@ -805,15 +801,15 @@ def learn_from_teacher_node(self, eager=False, canceller=None):
...
+            self.log.info(f"Teacher {str(current_teacher)} is perhaps down:{e}.")  # FIXME: This was printing the node bytestring. Is this really necessary?  #1712
```
Comment (derekpierre) [inline nucypher/network/nodes.py:805]: This logging is painful when it happens - can we do checksum address or `current_teacher.seed_node_metadata(as_teacher_uri=True)` or something simple like that. If needed, can we hit the `/public_information` endpoint after the fact to get the node bytes.

File: nucypher/network/nodes.py (hunk 1 above)
Comment (KPrasch) [reply nucypher/network/nodes.py]: debugging artifact - fixed.

File: nucypher/network/nodes.py (hunk 2 above)
Comment (KPrasch) [reply nucypher/network/nodes.py]: ✔️

File: nucypher/characters/lawful.py [hunk 3]
```diff
@@ -572,6 +573,7 @@ def retrieve(
 
         if not publisher_verifying_key:
             publisher_verifying_key = alice_verifying_key
+        publisher_verifying_key = PublicKey.from_bytes(bytes(publisher_verifying_key))
```
Comment (fjarri) [inline nucypher/characters/lawful.py:576]: I remember removing a line like that :) How do bytes turn up so deep in the call stack?

Comment (KPrasch) [reply nucypher/characters/lawful.py:576]: Heh - I had the same thought.  I think it's coming from here <https://github.com/nucypher/nucypher/blob/main/examples/finnegans_wake_demo/finnegans-wake-demo-testnet-l2.py#L106>

File: examples/finnegans_wake_demo/finnegans-wake-demo-testnet-l2.py [hunk 4]
```diff
@@ -114,17 +119,26 @@
...
+    verifying_key=verifying_key,
+)
 
 # These are the policy details for bob.
 # In this example bob will be granted access for 1 day,
 # trusting 2 of 3 nodes paying each of them 50 gwei per period.
 expiration = maya.now() + datetime.timedelta(days=1)
-threshold, shares = 2, 3
-price = alice.payment_method.quote(expiration=expiration, shares=shares).value
+threshold, shares = 1, 2
```
Comment (piotr-roslaniec) [inline examples/finnegans_wake_demo/finnegans-wake-demo-testnet-l2.py:131]: Is there any issue with Ursula's availability on Ibex?

File: nucypher/network/nodes.py [hunk 5]
```diff
@@ -805,15 +802,14 @@ def learn_from_teacher_node(self, eager=False, canceller=None):
         # These except clauses apply to the current_teacher itself, not the learned-about nodes.
         except NodeSeemsToBeDown as e:
             unresponsive_nodes.add(current_teacher)
-            self.log.info(
-                f"Teacher {str(current_teacher)} is perhaps down:{e}.")  # FIXME: This was printing the node bytestring. Is this really necessary?  #1712
+            self.log.info(f"Teacher {current_teacher.seed_node_metadata(as_teacher_uri=True)} is perhaps down:{e}.")
```
Comment (derekpierre) [inline nucypher/network/nodes.py:805]:
```suggestion
            self.log.info(f"Teacher {current_teacher.seed_node_metadata(as_teacher_uri=True)} is perhaps down: {e}.")
```

**Review by derekpierre — COMMENTED** [review]
Also created PR into airship to improve consistency in the external ip utilities unit test - https://github.com/KPrasch/nucypher/pull/36.

File: nucypher/characters/lawful.py (hunk 3 above)
Comment (KPrasch) [reply nucypher/characters/lawful.py:576]: Let's revisit this issue in a future PR.  Other than some negligible inefficiency I don't think there's any negative impact of accepting bytes or a public key here.

File: tests/utils/config.py [hunk 6]
```diff
@@ -86,11 +87,12 @@ def make_alice_test_configuration(payment_provider: str = None,
...
-                                payment_network=payment_network)
+                                payment_network=payment_network,
+                                policy_registry=test_params['registry'])
     return config
 
 
 def make_bob_test_configuration(**assemble_kwargs) -> BobConfiguration:
     test_params = assemble(**assemble_kwargs)
-    config = BobConfiguration(**test_params)
+    config = BobConfiguration(**test_params, policy_registry=test_params['registry'])
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply tests/utils/config.py]: ✔️

File: tests/acceptance/cli/test_alice.py [hunk 7]
```diff
@@ -115,7 +115,9 @@ def test_initialize_alice_with_custom_configuration_root(custom_filepath, click_
     # Files and Directories
     assert custom_filepath.is_dir(), 'Configuration file does not exist'
     assert (custom_filepath / 'keystore').is_dir(), 'Keystore does not exist'
-    assert (custom_filepath / 'known_nodes').is_dir(), 'known_nodes directory does not exist'
+
+    # TODO: Only using in-memory node storage for now
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply tests/acceptance/cli/test_alice.py:119]: ✔️

File: nucypher/network/nodes.py (hunk 5 above)
Comment (KPrasch) [reply nucypher/network/nodes.py]: ✔️

File: nucypher/network/nodes.py [hunk 8]
```diff
@@ -1068,23 +1064,28 @@ def validate_worker(self, registry: BaseContractRegistry = None, eth_provider_ur
...
             if registry:
+
                 if not self._operator_is_bonded(registry=registry):  # <-- Blockchain CALL
                     message = f"Operator {self.operator_address} is not bonded to staking provider {self.checksum_address}"
                     self.log.debug(message)
                     raise self.UnbondedOperator(message)
 
                 if self._staking_provider_is_really_staking(registry=registry, eth_provider_uri=eth_provider_uri):  # <-- Blockchain CALL
-                    self.verified_worker = True
+                    self.log.info(f'Verified operator {self}')
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/network/nodes.py:1082]: Yep - was referring to `Ursula.__str__` here.  That does not include the operator address though...

File: nucypher/config/constants.py [hunk 9]
```diff
@@ -75,4 +75,4 @@
 NUCYPHER_EVENTS_THROTTLE_MAX_BLOCKS = 'NUCYPHER_EVENTS_THROTTLE_MAX_BLOCKS'
 
 # Probationary period
-END_OF_POLICIES_PROBATIONARY_PERIOD = MayaDT.from_iso8601('2022-3-17T23:59:59.0Z')
+END_OF_POLICIES_PROBATIONARY_PERIOD = MayaDT.from_iso8601('2022-6-16T23:59:59.0Z')
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/config/constants.py:78]: ✔️

**Comment by KPrasch** [comment]
🤠 Merging -- Thanks for the contributions!

**Comment by vepkenez** [comment]
YES!

**Comment by jdbertron** [comment]
Yay ! Nice work guys

**Pull request merged**